pkg/serviceaccount/metadata/metadata.go

```diff
@@ -0,0 +1,157 @@
+)
+
+// IssuerMetadataServer is an HTTP server for metadata of the KSA token issuer.
+type IssuerMetadataServer struct {
+	metadataJSON []byte
+	keysetJSON   []byte
+}
+
+// issuerMetadata provides a minimal subset of OIDC provider metadata:
+// https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata
+type issuerMetadata struct {
+	Issuer string `json:"issuer"` // REQUIRED in OIDC; meaningful to relying parties.
+	// AuthzEndpoint string   `json:"authorization_endpoint"`                // REQUIRED in OIDC; but useless to relying parties.
+	JWKSURI       string   `json:"jwks_uri"`                              // REQUIRED in OIDC; meaningful to relying parties.
+	ResponseTypes []string `json:"response_types_supported"`              // REQUIRED in OIDC
+	SubjectTypes  []string `json:"subject_types_supported"`               // REQUIRED in OIDC
+	SigningAlgs   []string `json:"id_token_signing_alg_values_supported"` // REQUIRED in OIDC
+
+}
+
+// NewServer creates a new IssuerMetadataServer.
+// The issuer is the OIDC issuer; keys are the keys that may be used to sign
+// KSA tokens.
+func NewServer(iss, jwksURI string, keys []interface{}) (*IssuerMetadataServer, error) {
+	// Metadata and keys are expected to only change across restarts at present,
+	// so we just marshal immediately and serve the cached JSON bytes.
+	keyset, errs := serviceaccount.PublicJWKSFromKeys(keys)
+	if errs != nil {
+		return nil, errs
+	}
+
+	keysetJSON, err := json.Marshal(keyset)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal service account issuer JWKS: %v", err)
+	}
+
+	metadata := issuerMetadata{
+		Issuer:        iss,
+		JWKSURI:       jwksURI,
+		ResponseTypes: []string{"id_token"}, // Kubernetes only produces ID tokens
+		SubjectTypes:  []string{"public"},   // https://openid.net/specs/openid-connect-core-1_0.html#SubjectIDTypes
+		SigningAlgs:   getAlgs(keyset),      // REQUIRED by OIDC
+	}
+
+	metadataJSON, err := json.Marshal(metadata)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal service account issuer metadata: %v", err)
+	}
+
+	return &IssuerMetadataServer{
+		metadataJSON: metadataJSON,
+		keysetJSON:   keysetJSON,
+	}, nil
+}
+
+// Install adds this server to the request router c.
+func (s *IssuerMetadataServer) Install(c *restful.Container) {
+	// Container.Add "will detect duplicate root paths and exit in that case",
+	// so we need a root for /.well-known/openid-configuration to avoid conflicts.
+	cfg := new(restful.WebService)
+	cfg.Path(OpenIDConfigPath).Route(
+		cfg.GET("").
+			To(fromStandard(s.serveConfiguration)).
+			Doc("get serviceaccount issuer OIDC configuration").
+			Operation("getServiceAccountIssuerMetadata"))
+	c.Add(cfg)
+	// ...and another one for the JWKS
+	jwks := new(restful.WebService)
+	jwks.Path(JWKSPath).Route(
+		jwks.GET("").
+			To(fromStandard(s.serveKeys)).
+			Doc("get serviceaccount issuer keys").
+			Operation("getServiceAccountIssuerKeys"))
+	c.Add(jwks)
+}
```

Comment:
Could we untangle the go-restful and json-building bits? Something like:
* `pkg/routes` - contains go-restful bits, with one handler for jwks and one for the discovery doc... each handler only needs the bytes it will serve
* `pkg/serviceaccount` - contains logic to build the discovery doc and the keyset json based on keys/issuer/external address/jwksuri
* `pkg/master` - builds JSON with a call to `pkg/serviceaccount`, registers handlers from `pkg/routes`

Reply:
I'll give that a shot and see what happens.

Reply:
Looks cleaner now with that stuff in routes. Since the test is in a test package (kept it in `pkg/serviceaccount` as `package serviceaccount_test` so it can use the key helpers), no problem importing everything there and keeping the test the same.

Skipped doing anything to `pkg/master` for now (that logic is in `cmd/kube-apiserver/app/server.go` right now, not sure if we can access the data from `completedServerRunOptions` we need from `pkg/master`).

Reply:
> Skipped doing anything to pkg/master for now (that logic is in cmd/kube-apiserver/app/server.go right now, not sure if we can access the data from completedServerRunOptions we need from pkg/master).

You're already plumbing data from cmd/kube-apiserver/app/server to pkg/master in the form of an instantiated handler in ExtraConfig (`ServiceAccountIssuerMetadata *routes.IssuerMetadataServer`). If we instead put the discovery/keyset JSON blobs in ExtraConfig, you could construct/assign those in CreateKubeAPIServerConfig, and construct the route installer based on those blobs in pkg/master (alongside the other go-restful installers)

```go
	if utilfeature.DefaultFeatureGate.Enabled(features.ServiceAccountIssuerDiscovery) {
		if len(c.ExtraConfig.OpenIDMetadataJSON) > 0 && len(c.ExtraConfig.OpenIDKeysetJSON) > 0 {
			routes.NewIssuerMetadataServer(c.ExtraConfig.OpenIDMetadataJSON, c.ExtraConfig.OpenIDKeysetJSON).Install(s.Handler.GoRestfulContainer)
		}
	}
```

Reply:
That's a fair point, and still relatively simple. Will do.

Reply:
Done.